src/librustc/ty/sty.rs

```diff
@@ -1978,6 +1978,36 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
     }
 }
 
+#[derive(Copy, Clone, Debug, Hash, RustcEncodable, RustcDecodable, Eq, PartialEq, Ord, PartialOrd)]
+/// Used in the HIR by using `Unevaluated` everywhere and later normalizing to `Evaluated` if the
+/// code is monomorphic enough for that.
+pub enum LazyConst<'tcx> {
+    Unevaluated(DefId, &'tcx Substs<'tcx>),
+    Evaluated(&'tcx Const<'tcx>),
```

Comment:
Why is there another pointer indirection here? That's one more than before (and one more level of interning).